### ISSUE TITLE
[Feature] Show a warning when you set the DataContext from code

### ISSUE THREAD

**Issue by workgroupengineering** [AvaloniaUI/Avalonia#13387]
### Discussed in https://github.com/AvaloniaUI/Avalonia/discussions/13383

<div type='discussions-op-text'>

<sup>Originally posted by **FraserElectronics** October 25, 2023</sup>
I am struggling with a strange problem. I have created what amounts to a single page application with Avalonia. 
I have an About box that shares the same screen space as my main view controls but use the Z-Index to keep it behind the main view. I have the Z-Index of both the main view and About box view bound to properties on the main window view model.

The main view pseudo code:

```
	<Grid RowDefinitions="Auto">
		<Grid Grid.Row="0" ZIndex="{Binding MainViewZIndex}">
			<!-- All my main view controls here -->
		</Grid>
		<views:AboutView Grid.Row="0" ZIndex="{Binding AboutViewZIndex}"/>
	</Grid>
```

The AboutView is simply a user control.
During initial testing it worked great until I then bound a ViewModel to my AboutView user control.

So in my AboutView code-behind I have:

```
public partial class AboutView : UserControl
{
    public AboutView()
    {
        this.DataContext = new AboutViewViewModel();
        InitializeComponent();
    }
}
```

As soon as I run this and before anything is rendered, I get the exception:

```
Unable to cast object of type 'MyApp.Views.AboutView' to type 'MyApp.ViewModels.IMainWindowViewModel'.
```

If I remove the ```this.DataContext = ....``` from the AboutView, the application runs fine !

Anybody give me a clue as to what I am doing wrong?

Andy
</div>

## Describe the solution you'd like

Added the ability at analyzer to check whether a DataContext assignment is present in the code;

**Comment by timunie** [comment]
@workgroupengineering I don't see what the analyzer should do here? I mean it is a totally valid scenario to set / change DataContext from code. What happened here is, that x:DataType was not correct for the given DataContext.

**Comment by workgroupengineering** [comment]
> @workgroupengineering I don't see what the analyzer should do here? I mean it is a totally valid scenario to set / change DataContext from code. What happened here is, that x:DataType was not correct for the given DataContext.

The analyzer should display a warning that there may be expected unexpected, because the DataContext can be inherited from a parent element or set from Xaml.

**Comment by mgnslndh** [comment]
>  What happened here is, that x:DataType was not correct for the given DataContext. 

Wouldn't that be a check that the analyzer could do?

**Comment by robloo** [comment]
The only known case where the `DataContext` really shouldn't be set is in a templated control. And only `this.DataContext` should be avoided as it's valid to set child control's `DataContext` in that case as well.

In all other cases I can think of I've seen valid reasons to set the DataContext. It is a bit of a gotcha if you don't know what you are doing but I don't really know how an analyzer would be useful here. It would raise more false alarms than anything.

**Comment by maxkatz6** [comment]
Yep, I don't see how we could relyably add an analyzer here. It's a perfectly valid code.

A check for matching of `x:DataType` and `knownWindow.DataContext` is possible though.

**Comment by workgroupengineering** [comment]
> Yep, I don't see how we could relyably add an analyzer here. It's a perfectly valid code.

Did you see the linked discussion? When the developer is Avalonia and new or not from WPF, he may not know that a control can inherit the DataContex from its parent causing unwanted behavior.

The analizer should warn you if you use the this.DataContext or DataContex expression in the constructor.
